File: android/folkbot/src/main/java/kallaspriit/bot/Transport.java
package kallaspriit.bot;

public interface Transport {

  String getName();

  void send(String message);

  void addListener(Listener listener);

  enum State {
    DISCONNECTED,
    CONNECTING,
    CONNECTED
  }

  @SuppressWarnings("unused")
  interface Listener {

    default void onStateChanged(Transport transport, State newState, State previousState) {
    }

    default void onClientConnected(Transport transport) {
    }

    default void onError(Transport transport, Exception error) {
    }

    default void onMessageSent(Transport transport, String message) {
    }

    void onMessageReceived(Transport transport, String message);
  }
}

File: app-old/src/lib/transport/Transport.ts
export enum TransportState {
  DISCONNECTED = "DISCONNECTED",
  CONNECTING = "CONNECTING",
  RECONNECTING = "RECONNECTING",
  CONNECTED = "CONNECTED",
}

export interface TransportListener {
  onStateChanged(
    transport: Transport,
    newState: TransportState,
    previousState: TransportState
  ): void;
  onError(transport: Transport, error?: Error): void;
  onMessageSent(
    transport: Transport,
    message: string,
    wasSentSuccessfully: boolean
  ): void;
  onMessageReceived(transport: Transport, message: string): void;
}

export interface Transport {
  getName(): string;
  isAvailable(): boolean;
  getState(): TransportState;
  addListener(listener: TransportListener): void;
  connect(): Promise<void>;
  send(message: string): boolean;
}
